File: digital_wallet_nodejs/src/services/adminService.js
const pool = require('../config/db');
const AppError = require('../utils/AppError');

async function listUsers(search, page, size) {
  const clampedSize = Math.min(100, Math.max(1, size || 20));
  const offset = (Math.max(1, page || 1) - 1) * clampedSize;
  const searchPattern = `%${search || ''}%`;

  const countResult = await pool.query(
    'SELECT COUNT(*)::int AS total FROM users WHERE username ILIKE $1',
    [searchPattern]
  );

  const result = await pool.query(
    'SELECT id, username, role, created_at AS "createdAt" FROM users WHERE username ILIKE $1 ORDER BY id LIMIT $2 OFFSET $3',
    [searchPattern, clampedSize, offset]
  );

  return {
    data: result.rows.map(u => ({
      id: parseInt(u.id, 10),
      username: u.username,
      role: u.role,
      createdAt: u.createdAt,
    })),
    total: countResult.rows[0].total,
    page: Math.max(1, page || 1),
    size: clampedSize,
  };
}

async function getUserDetail(userId) {
  const userResult = await pool.query(
    'SELECT id, username, role, created_at AS "createdAt" FROM users WHERE id = $1',
    [userId]
  );
  const user = userResult.rows[0];
  if (!user) {
    throw new AppError(404, 'User not found: ' + userId);
  }

  const walletResult = await pool.query(
    'SELECT id, user_id AS "userId", currency, balance, version, updated_at AS "updatedAt" FROM wallets WHERE user_id = $1',
    [userId]
  );

  let transactions = [];
  if (walletResult.rows[0]) {
    const txResult = await pool.query(
      'SELECT id, from_wallet_id AS "fromWalletId", to_wallet_id AS "toWalletId", amount, tx_type AS "txType", status, created_at AS "createdAt" FROM transactions WHERE from_wallet_id = $1 OR to_wallet_id = $1 ORDER BY created_at DESC LIMIT 5',
      [walletResult.rows[0].id]
    );
    transactions = txResult.rows.map(tx => ({
      id: parseInt(tx.id, 10),
      fromWalletId: tx.fromWalletId ? parseInt(tx.fromWalletId, 10) : null,
      toWalletId: tx.toWalletId ? parseInt(tx.toWalletId, 10) : null,
      amount: parseFloat(tx.amount),
      txType: tx.txType,
      status: tx.status,
      createdAt: tx.createdAt,
    }));
  }

  return {
    id: parseInt(user.id, 10),
    username: user.username,
    role: user.role,
    createdAt: user.createdAt,
    wallet: walletResult.rows[0] ? {
      id: parseInt(walletResult.rows[0].id, 10),
      userId: parseInt(walletResult.rows[0].userId, 10),
      currency: walletResult.rows[0].currency,
      balance: parseFloat(walletResult.rows[0].balance),
      version: parseInt(walletResult.rows[0].version, 10),
      updatedAt: walletResult.rows[0].updatedAt,
    } : null,
    recentTransactions: transactions,
  };
}

async function disableUser(userId) {
  const check = await pool.query('SELECT id FROM users WHERE id = $1', [userId]);
  if (check.rows.length === 0) {
    throw new AppError(404, 'User not found: ' + userId);
  }
  await pool.query("UPDATE users SET role = 'ROLE_DISABLED' WHERE id = $1", [userId]);
}

async function enableUser(userId) {
  const check = await pool.query('SELECT id FROM users WHERE id = $1', [userId]);
  if (check.rows.length === 0) {
    throw new AppError(404, 'User not found: ' + userId);
  }
  await pool.query("UPDATE users SET role = 'ROLE_USER' WHERE id = $1", [userId]);
}

async function listTransactions(username, fromDate, toDate, page, size) {
  const clampedSize = Math.min(100, Math.max(1, size || 20));
  const offset = (Math.max(1, page || 1) - 1) * clampedSize;

  let conditions = [];
  let params = [];
  let paramIndex = 1;

  if (username) {
    conditions.push('(fu.username ILIKE $' + paramIndex + ' OR tu.username ILIKE $' + paramIndex + ')');
    params.push('%' + username + '%');
    paramIndex++;
  }

  if (fromDate) {
    conditions.push('t.created_at >= $' + paramIndex);
    params.push(fromDate);
    paramIndex++;
  }

  if (toDate) {
    conditions.push("t.created_at < $" + paramIndex + "::date + INTERVAL '1 day'");
    params.push(toDate);
    paramIndex++;
  }

  const whereClause = conditions.length > 0 ? 'WHERE ' + conditions.join(' AND ') : '';

  const countResult = await pool.query(
    'SELECT COUNT(*)::int AS total FROM transactions t LEFT JOIN wallets fw ON t.from_wallet_id = fw.id LEFT JOIN users fu ON fw.user_id = fu.id LEFT JOIN wallets tw ON t.to_wallet_id = tw.id LEFT JOIN users tu ON tw.user_id = tu.id ' + whereClause,
    params
  );

  const result = await pool.query(
    'SELECT t.id, t.from_wallet_id AS "fromWalletId", t.to_wallet_id AS "toWalletId", t.amount, t.tx_type AS "txType", t.status, t.created_at AS "createdAt", fu.username AS "fromUsername", tu.username AS "toUsername" FROM transactions t LEFT JOIN wallets fw ON t.from_wallet_id = fw.id LEFT JOIN users fu ON fw.user_id = fu.id LEFT JOIN wallets tw ON t.to_wallet_id = tw.id LEFT JOIN users tu ON tw.user_id = tu.id ' + whereClause + ' ORDER BY t.created_at DESC LIMIT $' + paramIndex + ' OFFSET $' + (paramIndex + 1),
    [...params, clampedSize, offset]
  );

  return {
    data: result.rows.map(tx => ({
      id: parseInt(tx.id, 10),
      fromWalletId: tx.fromWalletId ? parseInt(tx.fromWalletId, 10) : null,
      toWalletId: tx.toWalletId ? parseInt(tx.toWalletId, 10) : null,
      fromUsername: tx.fromUsername || null,
      toUsername: tx.toUsername || null,
      amount: parseFloat(tx.amount),
      txType: tx.txType,
      status: tx.status,
      createdAt: tx.createdAt,
    })),
    total: countResult.rows[0].total,
    page: Math.max(1, page || 1),
    size: clampedSize,
  };
}

async function getTransactionStats(fromDate, toDate) {
  let conditions = [];
  let params = [];

  if (fromDate) {
    conditions.push('created_at >= $' + (params.length + 1));
    params.push(fromDate);
  }
  if (toDate) {
    conditions.push("created_at < $" + (params.length + 1) + "::date + INTERVAL '1 day'");
    params.push(toDate);
  }

  const whereClause = conditions.length > 0 ? 'WHERE ' + conditions.join(' AND ') : '';

  const summaryResult = await pool.query(
    'SELECT COUNT(*)::bigint AS total, COALESCE(SUM(amount), 0) AS amount FROM transactions ' + whereClause,
    params
  );

  const detailResult = await pool.query(
    'SELECT DATE(created_at) as date, COUNT(*)::bigint as count, COALESCE(SUM(amount), 0) as amount FROM transactions ' + whereClause + ' GROUP BY DATE(created_at) ORDER BY date',
    params
  );

  return {
    totalTransactions: parseInt(summaryResult.rows[0].total, 10),
    totalAmount: parseFloat(summaryResult.rows[0].amount),
    dailyVolume: detailResult.rows.map(d => ({
      date: d.date,
      count: parseInt(d.count, 10),
      amount: parseFloat(d.amount),
    })),
  };
}

module.exports = { listUsers, getUserDetail, disableUser, enableUser, listTransactions, getTransactionStats };

File: digital_wallet_fastapi/app/services/admin_service.py
from datetime import date
from decimal import Decimal

from sqlalchemy import select, func, text, and_, or_
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import aliased

from app.exceptions.handlers import AppException
from app.models.user import User
from app.models.wallet import Wallet
from app.models.transaction import Transaction


async def list_users(session: AsyncSession, search: str, page: int, size: int) -> dict:
    page = max(page, 1)
    size = max(min(size, 100), 1)
    offset = (page - 1) * size

    if search:
        stmt = select(User).where(User.username.ilike(f"%{search}%"))
        count_stmt = select(func.count()).select_from(User).where(User.username.ilike(f"%{search}%"))
    else:
        stmt = select(User)
        count_stmt = select(func.count()).select_from(User)

    stmt = stmt.order_by(User.id).offset(offset).limit(size)
    result = await session.execute(stmt)
    users = result.scalars().all()

    count_result = await session.execute(count_stmt)
    total = count_result.scalar()

    return {
        "data": [{"id": u.id, "username": u.username, "role": u.role, "createdAt": u.created_at} for u in users],
        "page": page,
        "size": size,
        "total": total,
    }


async def get_user_detail(session: AsyncSession, user_id: int) -> dict:
    stmt = select(User).where(User.id == user_id)
    result = await session.execute(stmt)
    user = result.scalar_one_or_none()
    if user is None:
        raise AppException(404, f"User not found: {user_id}")

    stmt = select(Wallet).where(Wallet.user_id == user_id)
    result = await session.execute(stmt)
    wallet = result.scalar_one_or_none()

    wallet_dict = None
    recent_tx_list = []

    if wallet is not None:
        wallet_dict = {
            "id": wallet.id, "userId": wallet.user_id, "currency": wallet.currency,
            "balance": str(wallet.balance), "version": wallet.version,
            "updatedAt": wallet.updated_at.isoformat() if wallet.updated_at else None,
        }

        stmt = (
            select(Transaction)
            .where((Transaction.from_wallet_id == wallet.id) | (Transaction.to_wallet_id == wallet.id))
            .order_by(Transaction.created_at.desc())
            .limit(5)
        )
        result = await session.execute(stmt)
        txs = result.scalars().all()
        recent_tx_list = [{
            "id": tx.id, "fromWalletId": tx.from_wallet_id, "toWalletId": tx.to_wallet_id,
            "amount": str(tx.amount), "txType": tx.tx_type, "status": tx.status,
            "createdAt": tx.created_at.isoformat() if tx.created_at else None,
        } for tx in txs]

    return {
        "id": user.id, "username": user.username, "role": user.role,
        "createdAt": user.created_at,
        "wallet": wallet_dict, "recentTransactions": recent_tx_list,
    }


async def disable_user(session: AsyncSession, user_id: int) -> None:
    stmt = select(User).where(User.id == user_id)
    result = await session.execute(stmt)
    user = result.scalar_one_or_none()
    if user is None:
        raise AppException(404, f"User not found: {user_id}")
    user.role = "ROLE_DISABLED"
    await session.flush()


async def enable_user(session: AsyncSession, user_id: int) -> None:
    stmt = select(User).where(User.id == user_id)
    result = await session.execute(stmt)
    user = result.scalar_one_or_none()
    if user is None:
        raise AppException(404, f"User not found: {user_id}")
    user.role = "ROLE_USER"
    await session.flush()


async def list_transactions(session: AsyncSession, username: str | None, from_date: date | None, to_date: date | None, page: int, size: int) -> dict:
    page = max(page, 1)
    size = max(min(size, 100), 1)
    offset = (page - 1) * size

    FromWallet = aliased(Wallet)
    ToWallet = aliased(Wallet)
    FromUser = aliased(User)
    ToUser = aliased(User)

    cols = (
        Transaction.id, Transaction.from_wallet_id, Transaction.to_wallet_id,
        Transaction.amount, Transaction.tx_type, Transaction.status, Transaction.created_at,
        FromUser.username.label("from_username"), ToUser.username.label("to_username"),
    )

    stmt = (
        select(*cols).select_from(Transaction)
        .outerjoin(FromWallet, Transaction.from_wallet_id == FromWallet.id)
        .outerjoin(ToWallet, Transaction.to_wallet_id == ToWallet.id)
        .outerjoin(FromUser, FromWallet.user_id == FromUser.id)
        .outerjoin(ToUser, ToWallet.user_id == ToUser.id)
    )

    conditions = []
    if username:
        conditions.append(or_(FromUser.username == username, ToUser.username == username))
    if from_date:
        conditions.append(Transaction.created_at >= from_date)
    if to_date:
        conditions.append(Transaction.created_at < to_date)

    if conditions:
        combined = and_(*conditions)
        stmt = stmt.where(combined)

    stmt = stmt.order_by(Transaction.created_at.desc()).offset(offset).limit(size)
    result = await session.execute(stmt)
    rows = result.all()

    # count
    count_stmt = (
        select(func.count()).select_from(Transaction)
        .outerjoin(FromWallet, Transaction.from_wallet_id == FromWallet.id)
        .outerjoin(ToWallet, Transaction.to_wallet_id == ToWallet.id)
        .outerjoin(FromUser, FromWallet.user_id == FromUser.id)
        .outerjoin(ToUser, ToWallet.user_id == ToUser.id)
    )
    if conditions:
        count_stmt = count_stmt.where(and_(*conditions))

    count_result = await session.execute(count_stmt)
    total = count_result.scalar()

    data = [{
        "id": row.id, "fromWalletId": row.from_wallet_id, "toWalletId": row.to_wallet_id,
        "amount": row.amount, "txType": row.tx_type, "status": row.status,
        "createdAt": row.created_at,
        "fromUsername": row.from_username, "toUsername": row.to_username,
    } for row in rows]

    return {"data": data, "page": page, "size": size, "total": total}


async def get_transaction_stats(session: AsyncSession, from_date: date | None, to_date: date | None) -> dict:
    conditions = []
    params = {}

    if from_date:
        conditions.append("created_at >= :from_date")
        params["from_date"] = from_date
    if to_date:
        conditions.append("created_at < :to_date")
        params["to_date"] = to_date

    where_clause = "WHERE " + " AND ".join(conditions) if conditions else ""

    count_sql = f"SELECT COUNT(*)::bigint FROM transactions {where_clause}"
    amount_sql = f"SELECT COALESCE(SUM(amount), 0) FROM transactions {where_clause}"
    daily_sql = f"SELECT DATE(created_at) AS date, COUNT(*)::bigint AS count, COALESCE(SUM(amount), 0) AS amount FROM transactions {where_clause} GROUP BY DATE(created_at) ORDER BY date"

    total_tx = (await session.execute(text(count_sql), params)).scalar()
    total_amt = (await session.execute(text(amount_sql), params)).scalar()
    daily_result = await session.execute(text(daily_sql), params)
    daily_rows = daily_result.all()

    return {
        "totalTransactions": total_tx,
        "totalAmount": total_amt,
        "dailyVolume": [{"date": str(r.date), "count": r.count, "amount": r.amount} for r in daily_rows],
    }
